server/genquery2/src/genquery2_sql.cpp

```diff
@@ -344,20 +344,39 @@ namespace
-                               iter->second);
+        if (_opts.admin_mode) {
+            if (const auto iter = _state.table_aliases.find("R_DATA_MAIN"); iter != std::end(_state.table_aliases)) {
+                sql += fmt::format(" left join R_OBJT_ACCESS pdoa on {}.data_id = pdoa.object_id"
+                                   " left join R_TOKN_MAIN pdt on pdoa.access_type_id = pdt.token_id"
+                                   " left join R_USER_GROUP pdug on pdoa.user_id = pdug.group_user_id"
+                                   " left join R_USER_MAIN pdu on pdug.user_id = pdu.user_id",
+                                   iter->second);
+            }
+
+            if (const auto iter = _state.table_aliases.find("R_COLL_MAIN"); iter != std::end(_state.table_aliases)) {
+                sql += fmt::format(" left join R_OBJT_ACCESS pcoa on {}.coll_id = pcoa.object_id"
+                                   " left join R_TOKN_MAIN pct on pcoa.access_type_id = pct.token_id"
+                                   " left join R_USER_GROUP pcug on pcoa.user_id = pcug.group_user_id"
+                                   " left join R_USER_MAIN pcu on pcug.user_id = pcu.user_id",
+                                   iter->second);
+            }
         }
+        else {
+            if (const auto iter = _state.table_aliases.find("R_DATA_MAIN"); iter != std::end(_state.table_aliases)) {
+                sql += fmt::format(" inner join R_OBJT_ACCESS pdoa on {}.data_id = pdoa.object_id"
+                                   " inner join R_TOKN_MAIN pdt on pdoa.access_type_id = pdt.token_id"
+                                   " inner join R_USER_GROUP pdug on pdoa.user_id = pdug.group_user_id"
+                                   " inner join R_USER_MAIN pdu on pdug.user_id = pdu.user_id",
+                                   iter->second);
+            }
 
-        if (const auto iter = _table_aliases.find("R_COLL_MAIN"); iter != std::end(_table_aliases)) {
-            sql += fmt::format(" inner join R_OBJT_ACCESS pcoa on {}.coll_id = pcoa.object_id"
-                               " inner join R_TOKN_MAIN pct on pcoa.access_type_id = pct.token_id"
-                               " inner join R_USER_GROUP pcug on pcoa.user_id = pcug.group_user_id"
-                               " inner join R_USER_MAIN pcu on pcug.user_id = pcu.user_id",
-                               iter->second);
+            if (const auto iter = _state.table_aliases.find("R_COLL_MAIN"); iter != std::end(_state.table_aliases)) {
+                sql += fmt::format(" inner join R_OBJT_ACCESS pcoa on {}.coll_id = pcoa.object_id"
+                                   " inner join R_TOKN_MAIN pct on pcoa.access_type_id = pct.token_id"
+                                   " inner join R_USER_GROUP pcug on pcoa.user_id = pcug.group_user_id"
+                                   " inner join R_USER_MAIN pcu on pcug.user_id = pcu.user_id",
+                                   iter->second);
```

Comment:
Are inner-joins really necessary?
Do left-joins solve the problem for all cases?

Reply:
That's what I thought would be the case, but benchtesting this produced the same results as the inner-joins.

More testing is needed to be fully sure. I think left-joins work (i.e. do not return all rows) because the non-admin queries always have a permission check condition attached to them. The table representing the right side of the join will contain null for the columns which are not matched and those nulls will always fail the condition check resulting in them being discarded from the results.

Reply:
I'm going to move forward with left and inner joins. It's not clear to me what the blast radius is for using left-joins. Limiting the use of left-joins (for this case) to rodsadmins is safer too.

I'll leave a comment highlighting this thought.